src/main/java/org/tugraz/sysds/runtime/lineage/LineageCache.java

```diff
@@ -1,60 +1,307 @@
 				MatrixObject mo = ec.getMatrixObject(((ComputationCPInstruction) inst).output);
-				LineageCache._cache.put(item, mo.acquireReadAndRelease());
+				LineageCache.put(inst, item, mo.acquireRead());
+				mo.release();
 			}
 		}
 	}
 	
-	public static boolean probe(LineageItem key) {
-		return _cache.containsKey(key);
+	public static void put(Instruction inst, LineageItem key, MatrixBlock value) {
+		if (_cache.containsKey(key)) {
+			// Read from cache, replace value and place the entry at the head of the linkedlist.
+			Entry oldItem = _cache.get(key);
+			oldItem._value = value;
+			delete(oldItem);
+			setHead(oldItem);
+		}
+		else {
+			// Create a new entry.
+			Entry newItem = new Entry(key, value);
+			if (isBelowThreshold(value)) 
+				// Place the entry at head position.
+				setHead(newItem);
+			else {
+				// Make space by removing or spilling LRU entries.
+				makeSpace(inst, value);
+				setHead(newItem);
+			}
+			_cache.put(key, newItem);	
+			if (DMLScript.STATISTICS)
+				LineageCacheStatistics.incrementMemWrites();
+			updateSize(value, true);
+		}
 	}
 	
-	public static MatrixBlock get(LineageItem key) {
-		return _cache.get(key);
+	public static boolean probe(LineageItem key) {
+		boolean p = (_cache.containsKey(key) || _spillList.containsKey(key));
+
+		if (!p && DMLScript.STATISTICS && _removelist.contains(key))
+			// The sought entry was in cache but removed later 
+			LineageCacheStatistics.incrementDelHits();
+
+		return p;
 	}
 	
 	public static void resetCache() {
 		_cache.clear();
+		_spillList.clear();
+		if (DMLScript.STATISTICS)
+			_removelist.clear();
 	}
 	
 	public static boolean reuse(Instruction inst, ExecutionContext ec) {
-		if (!DMLScript.LINEAGE && DMLScript.LINEAGE_REUSE)
+		if (!DMLScript.LINEAGE_REUSE)
 			return false;
 		
-		if (inst instanceof ComputationCPInstruction) {
+		if (inst instanceof ComputationCPInstruction &&
+			inst.getOpcode().equalsIgnoreCase("tsmm")) {
 			boolean reused = true;
 			LineageItem[] items = ((ComputationCPInstruction) inst).getLineageItems();
 			for (LineageItem item : items) {
-				if (LineageCache.probe(item))
-					ec.setMatrixOutput(((ComputationCPInstruction) inst).output.getName(), LineageCache.get(item));
-				else
+				if (LineageCache.probe(item)) {
+					MatrixBlock d = LineageCache.get(inst, item);
+					ec.setMatrixOutput(((ComputationCPInstruction) inst).output.getName(), d);
+				} else
 					reused = false;
 			}
 			return reused && items.length > 0;
-		} else {
+		} else 
 			return false;
+	}
+
+	public static MatrixBlock get(Instruction inst, LineageItem key) {
+		// This method is called only when entry is present either in cache or in local FS.
+		if (_cache.containsKey(key))
+		{
+			// Read and put the entry at head.
+			Entry e = _cache.get(key);
+			delete(e);
+			setHead(e);
+			if (DMLScript.STATISTICS)
+				LineageCacheStatistics.incrementMemHits();
+			return e._value;
+		}
+		else
+			return readFromLocalFS(inst, key);
```

Comment:
Doesn't this call fail if the key was never evicted? Unless I'm missing something you might need to probe the spill list first.

Reply:
As discussed, execution will reach here only when spilllist contains this item.